lib/ddtrace/contrib/sinatra/header_tagger_middleware.rb

```diff
@@ -0,0 +1,73 @@
+require 'ddtrace/contrib/sinatra/request_span'
+
+module Datadog
+  module Contrib
+    module Sinatra
+      # Middleware used for automatically tagging configured headers and handle request span
+      class HeaderTaggerMiddleware
+        def initialize(app)
+          @app = app
+        end
+
+        def call(env)
+          span = RequestSpan.span!(env)
+
+          # Request headers
+          parse_request_headers(env).each do |name, value|
+            span.set_tag(name, value) if span.get_tag(name).nil?
+          end
+
+          status, headers, response_body = @app.call(env)
+
+          # Response headers
+          parse_response_headers(headers).each do |name, value|
+            span.set_tag(name, value) if span.get_tag(name).nil?
+          end
+
+          span.finish
```

Comment:
I think this'll need to be in an `ensure` block, and only call the span if it isn't finished. Otherwise if the application raises an error, we could have an unclosed span which would cause the trace to disappear. We should also add a test to verify this doesn't happen (if we don't already have one.)